active/0031-new-monitor-histogram.md

```diff
@@ -0,0 +1,83 @@
+# Improve Monitor Histogram
+
+## Changelog
+
+* 2024-09-24: @zmstone Initial draft
+
+## Abstract
+
+This EIP proposes to improve the monitor histogram implementation to make it more compact and efficient.
+
+## Motivation
+
+The current (as of 5.8) monitor histogram has below issues:
+
+- It stores too much data but in the end only needed 1000 data points to return.
+- The talbe is of type `set` making it impossible to perform in-place compaction.
+- The data points are queried from clustered nodes using RPC calls, merged by the local node, sorted, then downsampled to 1000 records.
+- The data points are formated as JSON array with repeated field names.
+
+All these issues make the current implementation inefficient and slow and reflected as timeouts in the dashboard.
+
+## Design
+
+- Create a new `ordered_set` table, key being `{Timestamp, node()}` and value being the counters and gauges in a map.
+- Do not use `local_content` for the table, so there is no need to RPC when querying the data.
+- For counters, store absolute values instead of calculating deltas. This makes compaction easier: simply downsample the data points.
+- Downsample the data points to less than 1000 points in the table. See downsampling algorithm below.
+- Return more compact data format in the API response. See API response format below.
+
+### Downsampling Algorithm
+
+The downsampling algorithm is as follows:
+
+- For the last 1 hour, keep 1 data point per 10 seconds, (60 * 6 = 360) data points in total.
+- For the last 1 day, keep 1 data point per 10 minutes,  (23 * 6 = 198) data points in total.
+- For the last 1 week, keep 1 data point per 1 hour, (6 * 24 = 144) data points in total.
+
+The total number of downsampling data points should not exceed 700.
+
+When downsampling, the data point which is earlier but closest to a wall clock aligned interval should be kept.
+
+For example:
+- `12:35:00` when interval is 10s.
+- `12:40:00` when interval is 10m.
+- `13:00:00` when interval is 1h.
```

Comment:
I don't understand this example.  Maybe provide more context and the hypothetical current time?

Reply:
updated.